chia/util/file_keyring.py

```diff
@@ -0,0 +1,438 @@
+            passphrase = obtain_current_passphrase(use_passphrase_cache=True)
+        except Exception as e:
+            print(f"Unable to unlock the keyring: {e}")
+            sys.exit(1)
+
+        return FileKeyring.symmetric_key_from_passphrase(passphrase, salt)
+
+    def encrypt_data(self, input_data: bytes, key: bytes, nonce: bytes) -> bytes:
+        encryptor = ChaCha20Poly1305(key)
+        data = encryptor.encrypt(nonce, input_data, None)
+        return data
+
+    def decrypt_data(self, input_data: bytes, key: bytes, nonce: bytes) -> bytes:
+        decryptor = ChaCha20Poly1305(key)
+        output = decryptor.decrypt(nonce, input_data, None)
+        return output
+
+    def load_outer_payload(self):
+        if not self.keyring_path.is_file():
+            raise ValueError("Keyring file not found")
+
+        self.outer_payload_cache = dict(yaml.safe_load(open(self.keyring_path, "r")))
+        version = int(self.outer_payload_cache.get("version"))
+        if version > MAX_SUPPORTED_VERSION:
+            print(
+                f"Keyring format is unrecognized. Found version {version}"
+                ", expected a value <= {MAX_SUPPORTED_VERSION}"
+            )
+            print("Please update to a newer version")
+            sys.exit(1)
+
+        # Attempt to load the salt. It may not be present if the keyring is empty.
+        salt = self.outer_payload_cache.get("salt")
+        if salt:
+            self.salt = bytes.fromhex(salt)
+
+    def load_keyring(self, passphrase: str = None):
+        with self.load_keyring_lock:
+            self.needs_load_keyring = False
+
+        self.load_outer_payload()
+
+        # Missing the salt or nonce indicates that the keyring doesn't have any keys stored.
+        salt_str = self.outer_payload_cache.get("salt")
+        nonce_str = self.outer_payload_cache.get("nonce")
+        if not salt_str or not nonce_str:
+            return
+
+        salt = bytes.fromhex(salt_str)
+        nonce = bytes.fromhex(nonce_str)
+        key = None
+
+        if passphrase:
+            key = FileKeyring.symmetric_key_from_passphrase(passphrase, salt)
+        else:
+            key = FileKeyring.get_symmetric_key(salt)
+
+        encrypted_payload = base64.b64decode(yaml.safe_load(self.outer_payload_cache.get("data") or ""))
+        decrypted_data = self.decrypt_data(encrypted_payload, key, nonce)
+        if not self.have_valid_checkbytes(decrypted_data):
+            raise ValueError("decryption failure (checkbytes)")
+        inner_payload = decrypted_data[len(CHECKBYTES_VALUE) :]
+
+        self.payload_cache = dict(yaml.safe_load(inner_payload))
+
+    def is_first_write(self):
+        return self.outer_payload_cache == FileKeyring.default_outer_payload()
+
+    def write_keyring(self, fresh_salt: bool = False):
+        from chia.util.keyring_wrapper import KeyringWrapper
+
+        inner_payload = self.payload_cache
+        inner_payload_yaml = yaml.safe_dump(inner_payload)
```

Comment:
I do not think we should encrypt and decrypt yaml. Is it guaranteed to be deterministic? We are depending on the yaml serialization format for the keychain to function properly.

Why don't we instead use Streamable (which is very simple, appending of each fixed sized field), or just a custom serialization for our use case?

Reply:
Can you explain the issue(s) you foresee with using YAML (or similar, like JSON)? I understand that there might be subtle differences in the way the serialization spits out the document (maybe whitespace differences, or different ordering of key/value pairs). The payload is read/written as a whole without any random access, and file-locking-based RD/WR locks are employed to guard against corrupting the data. I suppose a future change in the yaml module might cause problems?

I'm open to changing this if there's a need. Perhaps we can revisit after this PR lands? Since the feature is disabled, a change in the format shouldn't affect anyone other than a few of us internally.

Reply:
Usually when signing or hashing anything, you don't want the payload being signed to change due to different library versions etc. For example signing JSON is a bad idea because it's not deterministic so signing on two versions will produce an invalid result.

I suppose when you are encrypting it's not a big deal because it's OK if the stuff being encrypted changes (and actually you force it to change on every write, right)?

That said, for debugging purposes it does not seem great that the encrypted payload is not guaranteed to be the same, given the same input data. I would much rather use our already established Streamable implementation, or just a custom serialization that is guaranteed to be the same serialization for the input date. This would also be a simple change and would remove the necessity of migrating encrypted payload later (and also having two formats).

Keep in mind that other software will likely access the keys as well so having a well specified stable format for the keys file is important.

Reply:
I'll have to revisit this later. I attempted to switch to a Streamable class hierarchy and convert my existing dictionary into that format, but ran into errors before getting very far. I'm going to pursue the remaining test failures for now.